perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_def_mfaEnforceSettings.java

```diff
@@ -0,0 +1,188 @@
+			JsonNode root = mapper.readTree(val);
+
+			// Check "all"
+			if (root.has("all") && root.get("all").asBoolean() && root.size() == 1) return;
+
+			int validNodes = 0;
+			// Check "include_categories"
+			if (root.has("include_categories")) {
+				isStringArrayNode(root.get("include_categories"), "include_categories");
+				validNodes += 1;
+
+				// Check "exclude_rps"
+				if (root.has("exclude_rps")) {
+					isStringArrayNode(root.get("exclude_rps"), "exclude_rps");
+					validNodes += 1;
+				}
+			}
+
+			// Check for NO additional nodes
+			if (root.size() == validNodes) return;
+
+			throw new WrongAttributeValueException(
+				"Attribute value " + val + " has incorrect format." +
+				" Allowed values are:" +
+				" empty string or null," +
+				" {\"all\":true}," +
+				" {\"include_categories\":[\"str1\",\"str2\"]}," +
+				" {\"include_categories\":[\"str1\",\"str2\"],\"exclude_rps\":[\"rp1\",\"rp2\"]}");
+		} catch (WrongAttributeValueException e) {
+			throw e;
+		} catch (Exception e) {
+			throw new WrongAttributeValueException("Attribute value " + val + " is not a valid JSON.");
+		}
+	}
+
+	/**
+	 * Checks that node is an array and all values are strings
+	 *
+	 * @param node JsonNode
+	 * @throws WrongAttributeValueException
+	 */
+	private void isStringArrayNode(JsonNode node, String name) throws WrongAttributeValueException {
+		// Check property is valid array
+		if (node.isArray()) {
+			// Check all items of array are string like
+			for (Iterator<JsonNode> it = node.elements(); it.hasNext(); ) {
+				JsonNode value = it.next();
+				if (!value.isTextual()) {
+					throw new WrongAttributeValueException("Property '" + name + "' has non textual value " + value);
+				}
+			}
+		} else {
+			throw new WrongAttributeValueException("Property '" + name + "' is not an array.");
+		}
+	}
+	
+	/**
+	 * The following restrictions are placed on the attribute value:
+	 * {"include_categories":["str1","str2"]} str1, str2 is an existing key in the entityless attribute mfaCategories
+	 * {"include_categories":["str1","str2"],"exclude_rps":["rp1","rp2"]} str1, str2 is an existing key in the entityless attribute mfaCategories and rp1, rp2 must exist inside the category
+	 *
+	 * @param perunSession PerunSession
+	 * @param user User
+	 * @param attribute Attribute of the user.
+	 *
+	 * @throws WrongReferenceAttributeValueException
+	 * @throws WrongAttributeAssignmentException
+	 */
+	@Override
+	public void checkAttributeSemantics(PerunSessionImpl perunSession, User user, Attribute attribute) throws WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		String val = attribute.valueAsString();
+		if (val == null || val.isEmpty()) return;
+
+		Set<String> includeCategories = null;
+		Set<String> excludeRps = null;
+		try {
+			final ObjectMapper mapper = new ObjectMapper();
+			JsonNode root = mapper.readTree(val);
+
+			// Syntax check ensures correctness for 'all'
+			if (root.has("all")) return;
```

Comment:
Nitpick - the check for `true` should theoretically be here since we're checking the semantics